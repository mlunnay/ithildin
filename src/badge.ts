import * as m from 'mithril';

/**
 * A material badge.
 * 
 * ## Attributes
 * | attribute | type    | default   | description |
 * |-----------|---------|-----------|------------|
 * | icon      | boolean | false     | Treats the child element as a material icon definition |
 * | overlap   | boolean | false     | Make the badge overlap with its container |
 * | nobackground | boolean | false  | Applies open-circle effect to badge |
 * | data | string |   | Assigns string value to badge. This is transformed to data-badge attribute |
 * | data-badge | string |   | Assigns string value to badge |
 */
export class Badge implements m.ClassComponent<m.Attributes> {
    view(vnode: m.Vnode<m.Attributes, this>) {
        // add data-badge if it is not already in vnode.attrs and vnode.attrs.data is set.
        var attrs = (<any>Object).assign({}, vnode.attrs, !vnode.attrs['data-badge'] && vnode.attrs.data ? {'data-badge': vnode.attrs.data} : {});
        var view: string;
        if(attrs.icon) {
            view = 'div.material-icons.mdl-badge';
            delete attrs.icon;
        }
        else
            view = 'span.mdl-badge';
        if(attrs.overlap)
            view += '.mdl-badge--overlap';
        if(attrs.nobackground)
            view += '.mdl-badge--no-background';
        
        
        return m(view, attrs, vnode.children);
    }
}