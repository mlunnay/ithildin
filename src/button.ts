import * as m from 'mithril';
import 'material-design-lite';

/**
 * A materialize Button.
 * 
 * ## Attributes
 * | attribute | type    | default   | description |
 * |-----------|---------|-----------|------------|
 * | element   | string  | 'button'  | The base element type for the button |
 * | raised    | boolean | false     | makes the button raised |
 * | ripple    | boolean | true      | adds the ripple effect to the button |
 * | fab       | boolean | false     | makes the button a floating action button, mutually exclusive of *raised* and *icon* |
 * | minifab   | boolean | false     | makes the button a mini floating action button, mutually excluive of *raised* and *icon* |
 * | icon      | boolean | false     | Applies icon (small plain circular) display effect, mutually exclusive of *raised*, *fab* and *minifab* |
 * | color     | string  | undefined | this makes the button colored. Valide values are:</br> colored: applies colored display effect (primary or accent color, depending on the type of button)</br> primary: applies primary color display effect</br> accent: applies accent color display effect |
 */
export class Button implements m.ClassComponent<m.Attributes> {
    oncreate(vnode: m.VnodeDOM<m.Attributes, this>) {
        componentHandler.upgradeElement(<HTMLElement>vnode.dom);
    }

    view(vnode: m.Vnode<m.Attributes, this>) {
        var classes = '.mdl-button.mdl-js-button';
        if(vnode.attrs.raised)
            classes += '.mdl-button--raised';
        if(vnode.attrs.ripple || vnode.attrs.ripple === undefined)
            classes += '.mdl-js-ripple-effect';
        if(vnode.attrs.fab && !vnode.attrs.raised)
            classes += '.mdl-button--fab';
        if(vnode.attrs.minifab && !vnode.attrs.raised) {
            if(vnode.attrs.fab === undefined)
                classes += '.mdl-button--fab';
            classes += '.mdl-button--mini-fab';
        }
        if(vnode.attrs.icon && !vnode.attrs.raised && !(vnode.attrs.fab || vnode.attrs.minifab))
            classes += '.mdl-button--icon';
        if(vnode.attrs.color === 'colored')
            classes += '.mdl-button--colored';
        if(vnode.attrs.color === 'primary')
            classes += '.mdl-button--primary';
        if(vnode.attrs.color === 'accent')
            classes += '.mdl-button--accent';
        
        return m((vnode.attrs.element || 'button') + classes, vnode.attrs, vnode.children);
    }
} 