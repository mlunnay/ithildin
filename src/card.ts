import * as m from 'mithril';

/**
 * A materialize Card representing a virtual piece of paper that contains related data — such as a photo, some text, and a link — that are all about a single subject.
 * 
 * ## Attributes
 * | attribute | type    | default   | description |
 * |-----------|---------|-----------|------------|
 * | shadoe    | number  |           | Assigns variable shadow depths (2, 3, 4, 6, 8, or 16) to card |
 * | title     | object or string |       | Adds a title section to the card |
 * | title.expand | boolean | false  | Expands the title section to the width of the card |
 * | title.border | boolean | false  | Adds a border to the title section |
 * | title.content | string or vnode(s) |   | The content of the title section |
 * | media     | object or vnode(s) |   | Adds a media section to the card |
 * | media.expand | boolean | false  | Expands the media section to the width of the card |
 * | media.border | boolean | false  | Adds a border to the media section |
 * | media.content | vnode(s) |   | The content of the media section |
 * | supportingText     | object or vnode(s) |   | Adds a supporting text section to the card |
 * | supportingText.expand | boolean | false  | Expands the supporting text section to the width of the card |
 * | supportingText.border | boolean | false  | Adds a border to the supporting text section |
 * | supportingText.content | vnode(s) |   | The content of the supporting text section |
 * | actions     | object or vnode(s) |   | Adds a media section to the card |
 * | actions.expand | boolean | false  | Expands the actions section to the width of the card |
 * | actions.border | boolean | false  | Adds a border to the actions section |
 * | actions.content | vnode(s) |   | The content of the actions section |
 */
export class Card implements m.ClassComponent<m.Attributes> {
    view(vnode: m.Vnode<m.Attributes, this>) {
        var children = [];

        var shadow = '';
        if([2, 3, 4, 6, 8, 16].indexOf(vnode.attrs.shadow) >= 0)
            shadow = '.mdl-shadow--' + vnode.attrs.shadow + 'dp';

        if(vnode.attrs.title) {
            let classes = '.mdl-card__title';
            if(vnode.attrs.title.expand)
                classes += '.mdl-card--expand';
            if(vnode.attrs.title.border)
                classes += '.mdl-card--border';
            let child = vnode.attrs.title.content;
            if(typeof vnode.attrs.title === 'string')
                child = m('h2.mdl-card__title-text', vnode.attrs.title);
            children.push(m(classes, child));
        }

        if(vnode.attrs.media) {
            let classes = '.mdl-card__media';
            if(vnode.attrs.media.border)
                classes += '.mdl-card--border'
            if(vnode.attrs.media.expand)
                classes += '.mdl-card--expand';
            let child = vnode.attrs.media.content;
            if(!vnode.attrs.media.content)
                child = vnode.attrs.media;
            children.push(m(classes, child));
        }

        if(vnode.attrs.supportingText) {
            let classes = '.mdl-card__supporting-text';
            if(vnode.attrs.supportingText.border)
                classes += '.mdl-card--border'
            if(vnode.attrs.supportingText.expand)
                classes += '.mdl-card--expand';
            let child = vnode.attrs.supportingText.content;
            if(!vnode.attrs.supportingText.content)
                child = vnode.attrs.supportingText;
            children.push(m(classes, child));
        }

        if(vnode.attrs.actions) {
            let classes = '.mdl-card__action';
            if(vnode.attrs.actions.border)
                classes += '.mdl-card--border'
            if(vnode.attrs.actions.expand)
                classes += '.mdl-card--expand';
            let child = vnode.attrs.actions.content;
            if(!vnode.attrs.actions.content)
                child = vnode.attrs.actions;
            children.push(m(classes, child));
        }

        return m('.mdl-card' + shadow, children);
    }
}