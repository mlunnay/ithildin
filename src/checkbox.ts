import * as m from 'mithril';
import 'material-design-lite';

/**
 * A materialize checkbox.
 * 
 * ## Attributes
 * | attribute | type  | default   | description |
 * |-----------|-------|-----------|------------|
 * | label     | string |          | The label for the checkbox |
 * | ripple    | boolean | true    | Applies ripple click effect |
 * | checked   | boolean | false   | Whether the checkbox is checked or not |
 */
export class Checkbox implements m.ClassComponent<m.Attributes> {
    static checkboxIdCounter = 0;
    id: string;
    
    oninit(vnode: any) {
        this.id = vnode.attrs.id ? vnode.attrs.id : '__checkbox__' + ++Checkbox.checkboxIdCounter;
    }

    oncreate(vnode: any) {
        componentHandler.upgradeElement(vnode.dom);
    }

    view(vnode: m.Vnode<m.Attributes, this>) {
        let { label, id, ripple, ...attrs } = vnode.attrs;

        return m(`label[for="${this.id}"].mdl-checkbox.mdl-js-checkbox` + (ripple || ripple === undefined ? '.mdl-js-ripple-effect' : ''), { className: ripple ? '' : '' },
            m(`input[type="checkbox"].mdl-checkbox__input#${this.id}`, attrs),
            m('span.mdl-checkbox__label', label))
    }
}
