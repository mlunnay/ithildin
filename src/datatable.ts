import * as m from 'mithril';
import 'material-design-lite';

/**
 * A materialize Data Table.
 * 
 * ## Attributes
 * | attribute | type  | default   | description |
 * |-----------|-------|-----------|------------|
 * | data      | tableData[] |     | The content of the data |
 * | columns   | {label, (numeric), (sort)}[] |     | The column definitions. |
 * | selectable | boolean | false  | Applies all/individual selectable behavior (checkboxes) |
 * | shadow   | number |           | Assigns variable shadow depths (2, 3, 4, 6, 8, or 16) to the data table |
 */
export class DataTable implements m.ClassComponent<m.Attributes> {
    private dom: HTMLElement;
    
    /**
     * returns a list indexes of selected rows
     */
    getSelected() {
        const selectedElements = this.dom.querySelectorAll('tbody > tr');
        var result: number[] = [];
        for(var i = 0; i < selectedElements.length; i++) {
            if(selectedElements[i].className.indexOf('is-selected') >= 0)
                result.push(i);
        }
        return result;
    }

    /**
     * Returns true if the given row is selected.
     * @param index the index of the row to check.
     */
    isSelected(index: number) {
        return this.dom.childElementCount > index && this.dom.children[index].className.indexOf('is-selected') >= 0;
    }

    oncreate(vnode: m.VnodeDOM<m.Attributes, any>) {
        this.dom = <HTMLElement>vnode.dom;
        componentHandler.upgradeElement(<HTMLElement>vnode.dom);
    }

    view(vnode: m.Vnode<m.Attributes, this>) {
        const { data, columns, shadow, selectable } = vnode.attrs;
        var classes: string[] = [];
        if([2, 3, 4, 6, 8, 16].indexOf(shadow) >= 0)
            classes.push('.mdl-shadow--' + shadow + 'dp');
        if(selectable)
            classes.push('.mdl-data-table--selectable') 

        return m('table.mdl-data-table.mdl-js-data-table' + classes.join(' '),
            m('thead',
                m('tr',
                    columns.map((column: any) => {
                        return m('th' + (column.numeric === false ? '.mdl-data-table__cell--non-numeric' : ''), column.label)
                    }))),
            m('tbody',
                data.map((row: any[]) =>
                m('tr',
                 row.map((data: any, i: number) => {
                    return m('td' + (vnode.attrs.columns[i].numeric === false ? '.mdl-data-table__cell--non-numeric' : ''), data)
                }))))
        );
    }
}