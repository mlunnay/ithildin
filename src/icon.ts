import * as m from 'mithril';

/**
 * A material icon.
 * pass the name of the material icon to display as its child.
 * 
 * * ### Example
 * ```javascript
 * m(Icon, 'autorenew');
 * ```
 * will display ![autorenew][https://storage.googleapis.com/material-icons/external-assets/v4/icons/svg/ic_autorenew_black_18px.svg]
 */
export class Icon implements m.ClassComponent<m.Attributes> {
    view(vnode: m.Vnode<m.Attributes, this>) {
        return m('i.material-icons', vnode.children);
    }
}