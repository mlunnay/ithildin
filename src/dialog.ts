import * as m from 'mithril';
import { Button } from './button';
import * as dialogPolyfill from 'dialog-polyfill';

/**
 * A materialize Dialog.
 * 
 * ## Attributes
 * | attribute | type  | default   | description |
 * |-----------|-------|-----------|------------|
 * | title     | string \| vnode |   | The title for the dialog |
 * | content   | string \| vnode |   | The main content for the dialog |
 * | actions   | string \| vnode |   | The action(s) for the dialog |
 * | fullWidthActions | boolean | false | Modifies the actions to each take the full width of the container. This makes each take their own line. |
 */
export class Dialog implements m.ClassComponent<m.Attributes> {
    private dom: { open: boolean,
            returnValue: string,
            show: () => void,
            showModal: () => void,
            close: (returnValue?: string) => void
        };

    show() {
        this.dom.show();
    }

    showModal() {
        this.dom.showModal();
    }

    close(returnValue?: string) {
        this.dom.close(returnValue);
    }

    get returnValue() { return this.dom.returnValue; }

    get isOpen() { return this.dom.open; }

    oncreate(vnode: m.VnodeDOM<m.Attributes, any>) {
        dialogPolyfill.registerDialog(vnode.dom);
    }

    view(vnode: m.Vnode<m.Attributes, this>) {
        var titleSection;
        if(vnode.attrs.title)
            titleSection = m('div.mdl-dialog__title', vnode.attrs.title);

        var actionSection;
        if (!vnode.attrs.actions)
            actionSection = m(Button, {onclick: () => this.close()}, 'Ok');
        else
            actionSection = vnode.attrs.actions
        
        return m('dialog.mdl-dialog',
            titleSection,
            m('div.mdl-dialog__content', vnode.attrs.content),
            m('div.mdl-dialog__actions' + (vnode.attrs.fullWidthActions ? '.mdl-dialog__actions--full-width' : ''), actionSection));
    }
}