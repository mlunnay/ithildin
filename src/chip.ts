import * as m from 'mithril';
import { Icon } from './icon';

/**
 * A materialize Chip component is a small, interactive element. Chips are commonly used for contacts, text, rules, icons, and photos.
 * 
 * ## Attributes
 * | attribute | type    | default   | description |
 * |-----------|---------|-----------|------------|
 * | element   | string  | 'span'    | The base element type for the chip |
 * | deletable | boolean | false     | Makes the chip deletable |
 * | ondelete  | function |          | A callback for when the delete button is clicked, this makes the delete element a button |
 * | deletelink | string |           | A link to call when the delete button is clicked, this makes the delete a link, mutually exclusive of ondelete |
 * | contact   | string \| vnode |   | Adds the node to the chips contact |
 */
export class Chip implements m.ClassComponent<m.Attributes> {
    view(vnode: m.Vnode<m.Attributes, this>) {
        var element;
        if(vnode.attrs.element) {
            element = vnode.attrs.element;
            delete vnode.attrs.element;
        }
        var outerClasses = '.mdl-chip';
        if (vnode.attrs.deletable)
            outerClasses += '.mdl-chip--deletable';

        var button;
        if (vnode.attrs.deletable && (vnode.attrs.ondelete || !vnode.attrs.deletelink)) {

            button = m('button.mdl-chip__action', vnode.attrs.ondelete ? { onclick: vnode.attrs.ondelete } : {}, m(Icon, 'cancel'));
        }
        else if (vnode.attrs.deletable && vnode.attrs.deletelink && !vnode.attrs.ondelete)
            button = m('a.mdl-chip__action', { href: vnode.attrs.deletelink }, m(Icon, 'cancel'))
        
        var contact;
        if(vnode.attrs.contact) {
            contact = vnode.attrs.contact;
            contact.attrs.class ? contact.attrs.class += ' mdl-chip__contact' : contact.attrs.class = 'mdl-chip__contact';
            delete vnode.attrs.contact;
        }

        return m((element || 'span') + outerClasses, vnode.attrs,
                contact,
                m('span.mdl-chip__text', vnode.children),
                button
            );
    }
}