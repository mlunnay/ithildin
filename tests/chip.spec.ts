import { } from 'jest';
import * as m from 'mithril';
import { tidyHtml } from './util/snapshots';
import { Chip } from '../src/chip';
import { MithrilQuery } from './util/mithrilQuery';

describe('Chip Component', () => {
    it('no options', () => {
        const cmp = new MithrilQuery(m(Chip, 'Basic chip'));
        expect(cmp).selectorToHave('span.mdl-chip span.mdl-chip__text:contains(Basic chip)', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
    it('button', () => {
        const cmp = new MithrilQuery(m(Chip, { element: 'button' }, 'Basic chip'));
        expect(cmp).selectorToHave('button.mdl-chip span.mdl-chip__text:contains(Basic chip)', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
    it('deletable (true)', () => {
        const cmp = new MithrilQuery(m(Chip, { deletable: true }, 'Basic chip'));
        expect(cmp).selectorToHave('span.mdl-chip.mdl-chip--deletable button.mdl-chip__action', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
    it('deletable (ondelete)', () => {
        const cmp = new MithrilQuery(m(Chip, { deletable: true, ondelete: () => alert('delete') }, 'Basic chip'));
        expect(cmp).selectorToHave('span.mdl-chip span.mdl-chip__text:contains(Basic chip)', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
    it('deletable (deletelink)', () => {
        const cmp = new MithrilQuery(m(Chip, { deletable: true, deletelink: '#' }, 'Basic chip'));
        expect(cmp).selectorToHave('a[href="#"].mdl-chip__action', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
    it('contact', () => {
        const cmp = new MithrilQuery(m(Chip, { contact: m('img', { src: "img.png" }) }, 'Contact chip'));
        expect(cmp).selectorToHave('img[src="img.png"].mdl-chip__contact', 1);
        const html = tidyHtml(cmp.toHtml(), { wrap: 0 });
        expect(html).toMatchSnapshot();
    });
});