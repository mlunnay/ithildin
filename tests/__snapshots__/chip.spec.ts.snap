// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip Component button 1`] = `"<button class=\\"mdl-chip\\"><span class=\\"mdl-chip__text\\">Basic chip</span></button>"`;

exports[`Chip Component contact 1`] = `"<span class=\\"mdl-chip\\"><img src=\\"img.png\\" class=\\"mdl-chip__contact\\"><span class=\\"mdl-chip__text\\">Contact chip</span></span>"`;

exports[`Chip Component deletable (deletelink) 1`] = `"<span deletable=\\"\\" deletelink=\\"#\\" class=\\"mdl-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Basic chip</span><a href=\\"#\\" class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></a></span>"`;

exports[`Chip Component deletable (ondelete) 1`] = `"<span deletable=\\"\\" class=\\"mdl-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Basic chip</span><button class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></button></span>"`;

exports[`Chip Component deletable (true) 1`] = `"<span deletable=\\"\\" class=\\"mdl-chip mdl-chip--deletable\\"><span class=\\"mdl-chip__text\\">Basic chip</span><button class=\\"mdl-chip__action\\"><i class=\\"material-icons\\">cancel</i></button></span>"`;

exports[`Chip Component no options 1`] = `"<span class=\\"mdl-chip\\"><span class=\\"mdl-chip__text\\">Basic chip</span></span>"`;
